lib/authentication.js

```diff
@@ -0,0 +1,172 @@
+'use strict';
+
+const MAX_PASS_LENGTH = 128;
+const MIN_PASS_LENGTH = 10;
+const MIN_PHRASE_LENGTH = 20;
+
+const MIN_LOGIN_LENGTH = 6;
+const MAX_LOGIN_LENGTH = 255;
+
+const passwordTests = {
+  MIN_LENGTH: [
+    password => password.length < MIN_PASS_LENGTH,
+    `The password must be longer than ${MIN_PASS_LENGTH} characters`
+  ],
+  MAX_LENGTH: [
+    password => password.length > MAX_PASS_LENGTH,
+    `The password must be less than ${MAX_PASS_LENGTH} characters`
+  ],
+  MIN_PHRASE_LENGTH: [
+    password => password.length < MIN_PHRASE_LENGTH,
+    `The passphrase must be longer than ${MIN_PHRASE_LENGTH} characters`
+  ],
+  REPEAT_CHARS: [
+    password => /(.)\1{2,}/.test(password),
+    'The password can not contain a sequence of repeated chars'
+  ],
+  ONE_LOWERCASE_CHAR: [
+    password => !/[a-z]/.test(password),
+    'The password must contain lowercase char'
+  ],
+  ONE_UPPERERCASE_CHAR: [
+    password => !/[A-Z]/.test(password),
+    'The password must contain uppercase char'
+  ],
+  ONE_NUMBER: [
+    password => !/[0-9]/.test(password),
+    'The password must contain number'
+  ],
+  ONE_SPECIAL_CHAR: [
+    password => !/[^A-Za-z0-9]/.test(password),
+    'The password must contain special char'
+  ]
+};
+
+const loginTests = {
+  MIN_LENGTH: [
+    login => login.length < MIN_LOGIN_LENGTH,
+    `The login must be longer than ${MIN_LOGIN_LENGTH} characters`
+  ],
+  MAX_LENGTH: [
+    login => login.length > MAX_LOGIN_LENGTH,
+    `The login must be less than ${MAX_LOGIN_LENGTH} characters`
+  ]
+};
+
+const loginPasswordTests = {
+  LOGIN_INCLUDES_PASSWORD: [
+    (login, password) => login.includes(password),
+    'The login can not contain a password'
+  ],
+  PASSWORD_INCLUDES_LOGIN: [
+    (login, password) => password.includes(login),
+    'The password can not contain a login'
+  ]
+};
+
+class AuthenticationStrength {
+  constructor(
+    hints, // Array, [{ required: Array, optional: Array }, ...]
+    bestPercent // Number, percent of best test group
+  ) {
+    this.hints = hints;
+    this.bestPercent = bestPercent;
+  }
+  get valid() {
+    return !!this.bestPercent;
+  }
+  get strength() {
+    if (this.bestPercent < 25) {
+      return 'Very weak';
+    } else if (this.bestPercent < 40) {
+      return 'Weak';
+    } else if (this.bestPercent < 60) {
+      return 'Good';
+    } else if (this.bestPercent < 80) {
+      return 'Strong';
+    } else {
+      return 'Very strong';
+    }
+  }
+}
+
+const makeTest = (
+  // Function that checks the arguments on a test suite
+  tests, // Array, array of password/login tests
+  testGroups, // Array, [ [[requiredTest, ...], [optionalTest, ...]], ...]
```

Comment:
I don't understand why is it possible to have multiple groups of required and optional tests? Isn't it one array of required tests and one of optional? Then it's possible to just use
```
{
  required: [],
  optional: [],
}
```

Reply:
I think that the explanation by the example will be more clear. The password will be valid if (the password is longer than the minimum password length + is less than the maximum password length + there are no duplicate characters) or if (the password is greater than MIN_PHRASE_LINGTX). Each of the required conditions has its own optional conditions.
```
const testGroups = [
  [
    ['MIN_LENGTH', 'MAX_LENGTH', 'REPEAT_CHARS'],
    ['ONE_LOWERCASE_CHAR', 'ONE_NUMBER', 'ONE_UPPERERCASE_CHAR']
  ], [
    ['MIN_PHRASE_LENGTH'],
    ['REPEAT_CHARS']
  ]
];
```

Reply:
Do I understand it correctly that for 'MIN_LENGTH' optional is 'ONE_LOWERCASE_CHAR' and so on?

Reply:
Yes.

Reply:
Could you elaborate, I proposed 2 ways.
Also, either way, I do not see why there should be such a distinction and how first and second subarrays are connected. Why can't just there be an object of 2 arrays as I suggested?

Reply:
If we use an object with two arrays, it will be difficult for us to build a disjunction inside each of the arrays. Another example: if the login is an email, then it is valid and does not have optional tests, but if the login is not an email, it can have optional tests.
```
const loginTestGroups = [
  [
    ['IS_EMAIL'],
    []
  ], [
    ['MIN_LENGTH', 'MAX_LENGTH'],
    ['IS_BRUTE_FORCE_VALUE', 'REPEAT_CHARS']
  ]
];
```
How to do this with two arrays?

Reply:
So you mean that all of the top array entries are a disjunction (it only has to pass a single of those)? What is the use case for this?

Reply:
Disjunction applies only to test groups (group is array of required tests and array of optional tests). Therefore, in order for the password to be valid it must pass all the required tests, at least one group. Use case: my first example with MIN_LENGTH and MIN_PHRASE_LENGTH. https://www.owasp.org/index.php/Authentication_Cheat_Sheet#Password_Length

Reply:
Then I'd still suggest to use {  required: [],  optional: [] } for subarrays to avoid overcomplicating the syntax.
So it will look something like
```js
[
  {
    required: ['IS_EMAIL'],
  },
  {
    required: ['MIN_LENGTH', 'MAX_LENGTH'],
    optional: ['IS_BRUTE_FORCE_VALUE', 'REPEAT_CHARS']
  }
]
```

Reply:
Ok, thanks. I also think that it will look better.